File: Programmers/Algorithm_Kit/06_Greedy/001.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

int solution(int n, vector<int> lost, vector<int> reserve) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    
    vector<int> ret(n, 1);
    
    for(int i : lost) {
        ret[i - 1]--;
    }
    
    for(int i : reserve) {
        ret[i - 1]++;
    }
    
    for(int i = 0; i < n; i++) {
        if(!ret[i]){
            if(i > 0 && ret[i - 1] == 2) {
                ret[i - 1]--;
                ret[i]++;
            }
            else if(i < n - 1 && ret[i + 1] == 2) {
                ret[i + 1]--;
                ret[i]++;
            }
        }
    }
    
    int answer = 0;
    
    for(int i : ret){
        if(i) {
            answer++;
        }
    }
    
    return answer;
}

File: programmers/algorithmkit/06_greedy/001.java
import java.util.*;

class Solution {
    public int solution(int n, int[] lost, int[] reserve) {
        int answer = 0;

        int[] temp = new int[n];

        Arrays.fill(temp, 1);

        for (int i : lost) {
            temp[i - 1]--;
        }

        for (int i : reserve) {
            temp[i - 1]++;
        }

        for (int i = 0; i < n; i++) {
            if (temp[i] == 0) {
                if (i > 0 && temp[i - 1] == 2) {
                    temp[i - 1]--;
                    temp[i]++;
                } else if (i < n - 1 && temp[i + 1] == 2) {
                    temp[i + 1]--;
                    temp[i]++;
                }
            }
        }

        for (int i : temp) {
            if (i > 0) {
                answer++;
            }
        }
        
        return answer;
    }
}
